Tune links and hashes

csillag opened the pull request:
@donouwens wrote:

> **TX details, mobile view:**
> we should use the space in the row more to minimise the chance of needing to truncate.

Up to now, on mobile, we were always truncating hashes to a fixed length (the same as in tables).

With this change, we can now display them adaptively, that is, taking up all the space that is available.

| Component |Before | After| Points of interest |
|  ---- | ----- | ---- | ---- |
| TX details on mobile | ![image](https://github.com/oasisprotocol/explorer/assets/2093792/eaeeb2a2-4348-495e-bbbb-84c8a122c483) |  ![image](https://github.com/oasisprotocol/explorer/assets/2093792/e57e6dd2-59b9-4416-b22d-cc97447cdcf2) | Fixed lost space and clashing tooltips for "From" and "To" |
| Account details on mobile | ![image](https://github.com/oasisprotocol/explorer/assets/2093792/1f15fe45-0983-4fad-8383-f3db69cc4742)  |  ![image](https://github.com/oasisprotocol/explorer/assets/2093792/1e0856cd-5470-4ba3-aa43-ed5d04487ab6)  | Fixed lost space for address  |

A similar change has also been implemented for block hash links and other similar things.

csillag commented:
Works with Chrome, but not with FF. Lemme fix that...

csillag commented:
Works with Chrome, but not with FF. Lemme fix that...

csillag commented:
> 1. When I view the preview on mobile it's still truncated ("0x1234...abcdef") - that' correct behaviour, right?

It was not. Now fixed.

Firefox: 

![image](https://github.com/oasisprotocol/explorer/assets/2093792/ca75a20e-414c-4bff-ba31-1b0a93c0be92)

Chrome:

![image](https://github.com/oasisprotocol/explorer/assets/2093792/ad3ebcd8-ef4e-4226-86c5-9485fd70246b)

> 2. Would we ever go over two lines on any (normal) screen (mobile/tablet/desktop)?

For account address, no, not before we bring in named accounts in another PR.

lukaw3d commented on src/app/components/AdaptiveTrimmer/AdaptiveTrimmer.tsx:
```diff
@@ -0,0 +1,140 @@
+  }, [initDiscovery])
+
+  useEffect(() => {
+    if (inDiscovery) {
+      if (!textRef.current) {
+        return
+      }
+      const isOverflow = textRef.current.scrollWidth > textRef.current.clientWidth
+
+      if (isOverflow) {
+        // This is too much
+
+        // Update known bad length
+        const newSmallestKnownBad = Math.min(currentLength, smallestKnownBad)
+        setSmallestKnownBad(newSmallestKnownBad)
+
+        // We should try something smaller
+        attemptShortenedContent(Math.floor((largestKnownGood + newSmallestKnownBad) / 2))
+      } else {
+        // This is OK
+
+        // Update known good length
+        const newLargestKnownGood = Math.max(currentLength, largestKnownGood)
+        setLargestKnownGood(currentLength)
+
+        if (currentLength === fullLength) {
+          // The whole thing fits, so we are good.
+          setInDiscovery(false)
+        } else {
+          if (currentLength + 1 === smallestKnownBad) {
+            // This the best we can do, for now
+            setInDiscovery(false)
+          } else {
+            // So far, so good, but we should try something longer
+            attemptShortenedContent(Math.floor((newLargestKnownGood + smallestKnownBad) / 2))
+          }
+        }
+      }
+    }
+  }, [inDiscovery, currentLength, largestKnownGood, smallestKnownBad, attemptShortenedContent, fullLength])
```
Too much complexity for a small improvement

lukaw3d replied:
If we want this, please implement it using CSS. That will also fix <kbd>CTRL+F</kbd> "0x123456789" finding "0x123...789"

e.g. https://github.com/oasisprotocol/oasis-wallet-web/compare/master...lw/abandoned/address-ellipsis

csillag replied:
A pure CSS solution would be tempting, if it can fulfill all the requirements. Things to check:

- Can we instruct CSS to shorten the string by adding eclipses either in the middle or the end, based on our specific preference?
- If CSS is doing the shortening, then how do we know that we need to add a tooltip?

lukaw3d replied:
- yes
- no

csillag replied:
Can the CSS solution do something similar to [this](https://a193f9df.oasis-explorer.pages.dev/testnet/pontusx/search?q=kunst)?

(When opened with narrow horizontal size, the name will be dynamically resized, beginning and end hidden, but it will always be displayed in a way that the highlight is within the visible section, even if it's not at the middle.)

csillag replied:
OK, going with css.

csillag commented:
This branch has too many merge conflicts now. I have rebased the code (on the Pontus-X branch), and extracted the relevant parts again into #1396. Closing this as obsolete.
